src/async.js

```diff
@@ -28,53 +30,87 @@ const POLL_INTERVAL_MILLISECONDS = 100;
  * NOTE: When using {@link setupAndTeardownApp}, it is not necessary to call this function separately.
  */
 export function setupAsync() {
-  beforeEach('create test promise', () => {
+  let timeoutHandle = null;
+
+  beforeEach('create test promise', function() {
     if (testPromise) {
       return;
     }
+
     testPromise = Promise.resolve();
+    currentErrorMessage = null;
+
+    const timeout = this.currentTest.timeout();
+    this.currentTest.timeout(0); // disable built-in timeout;
+
+    timeoutHandle = setTimeout(() => {
+      throw new Error(getErrorMessage());
+    }, timeout);
   });
 
-  afterEach('check test promise for errors', function() {
-    if (!testPromise) {
-      return;
+  afterEach('remove test promise', () => {
+    testPromise = null;
+    clearTimeout(timeoutHandle);
+  });
+}
+
+/**
+ * Drop-in replacement for the regular `it` function, with the same signature.
+ * This will augments the `it` functionality to automatically return the global test promise behind the scenes, to make
+ * the asynchronous helpers work as expected without any need for the developer to return a promise or call a `done` function.
+ * The `done` parameter can still be used and will override the global test promise flow as per the default mocha behavior.
+ *
+ * @NOTE: None of the async helpers can be used outside a function passed to `asyncIt`.
+ */
+export const asyncIt = createAsyncIt(it);
+asyncIt.only = createAsyncIt(it.only);
+asyncIt.skip = createAsyncIt(it.skip);
+
+function createAsyncIt(regularItFunction) {
+  return (description, func) => {
+    if (!func) {
+      regularItFunction(description);
+    } else {
+      regularItFunction(description, wrapForAsyncIt(func));
     }
+  };
+}
 
-    const testTimeout = this.currentTest.timeout();
-    this.timeout(0); // disable mocha timeout, since we're taking it over.
-
-    return new Promise((resolve, reject) => {
-      const timeoutHandle = setTimeout(() => {
-        const errorMessage = getErrorMessage();
-        cleanUp();
-        reject(new Error(errorMessage));
-      }, testTimeout);
-
-      testPromise
-        .then(returnValue => {
-          cleanUp();
-          resolve(returnValue);
-        })
-        .catch(error => {
-          cleanUp();
-          reject(error);
-        });
-
-      const cleanUp = () => {
-        clearTimeout(testPromise.timeoutHandle);
-        testPromise = null;
-        clearTimeout(timeoutHandle);
-        this.timeout(testTimeout);
-      };
-    });
-  });
+function wrapForAsyncIt(func) {
+  const wrappedBody = done => {
+    isInsideAsyncIt = true;
+    const funcResult = func(done);
+    isInsideAsyncIt = false;
+
+    if (isPromise(funcResult)) {
+      testPromise = testPromise.then(() => {
+        currentErrorMessage = 'Promise returned from `asyncIt` never resolved.';
+        return funcResult;
+      })
+    }
+
+    return testPromise;
+  };
+
+  const funcTakesDoneAsParameter = func.length === 1;
```

Comment:
I did not know `Function.length` stated how many params the function takes!

Reply:
It's an old trick I learned when working with ActionScript3 :-)